cmd/zfs/zfs_main.c

```diff
@@ -7289,6 +7389,230 @@ zfs_do_change_key(int argc, char **argv)
+	int ret = 0, c, op = 0;
+
+	if (argc < 3)
+		usage(B_FALSE);
+
+	while ((c = getopt(argc, argv, "p:cCsdkr0")) != -1) {
+		switch (c) {
+		case 'c':
+			if (op != ZFS_PROJECT_INVALID) {
+				(void) fprintf(stderr, gettext("cannot "
+				    "specify '-c' '-C' '-s' together\n"));
+				usage(B_FALSE);
+			}
+
+			op = ZFS_PROJECT_CHECK;
+			break;
+		case 'C':
+			if (op != ZFS_PROJECT_INVALID) {
+				(void) fprintf(stderr, gettext("cannot "
+				    "specify '-c' '-C' '-s' together\n"));
+				usage(B_FALSE);
+			}
+
+			op = ZFS_PROJECT_CLEAR;
+			break;
+		case 's':
+			if (op != ZFS_PROJECT_INVALID) {
+				(void) fprintf(stderr, gettext("cannot "
+				    "specify '-c' '-C' '-s' together\n"));
+				usage(B_FALSE);
+			}
+
+			op = ZFS_PROJECT_SET;
+			break;
+		case 'd':
+			if (recursive) {
+				(void) fprintf(stderr, gettext("cannot "
+				    "specify '-d' and '-r' together\n"));
+				usage(B_FALSE);
+			}
+
+			dironly = B_TRUE;
+			break;
+		case 'k':
+			keep = B_TRUE;
+			break;
+		case 'r':
+			if (dironly) {
+				(void) fprintf(stderr, gettext("cannot "
+				    "specify '-d' and '-r' together\n"));
+				usage(B_FALSE);
+			}
+
+			recursive = B_TRUE;
+			break;
+		case 'p':
+			projid = strtoull(optarg, NULL, 0);
+			if (projid == ZFS_INVALID_PROJID) {
+				(void) fprintf(stderr,
+				    gettext("invalid project ID\n"));
+				usage(B_FALSE);
+			}
+
+			break;
+		case '0':
+			newline = B_FALSE;
+			break;
+		default:
+			(void) fprintf(stderr, gettext("invalid option '%c'\n"),
+			    optopt);
+			usage(B_FALSE);
+		}
+	}
+
+	switch (op) {
+	case ZFS_PROJECT_CHECK:
+		if (!recursive) {
+			if (projid != ZFS_INVALID_PROJID) {
+				(void) fprintf(stderr, gettext(
+				    "'-p' need to be together with '-c -r'\n"));
+				usage(B_FALSE);
+			}
+			if (!newline) {
+				(void) fprintf(stderr, gettext(
+				    "'-0' need to be together with '-c -r'\n"));
+				usage(B_FALSE);
+			}
+		}
+		if (keep) {
+			(void) fprintf(stderr,
+			    gettext("'-k' is useless together with '-c'\n"));
+			usage(B_FALSE);
+		}
+		break;
+	case ZFS_PROJECT_CLEAR:
+		if (!newline) {
+			(void) fprintf(stderr,
+			    gettext("'-0' is useless together with '-C'\n"));
+			usage(B_FALSE);
+		}
+		if (dironly) {
+			(void) fprintf(stderr,
+			    gettext("'-d' is useless together with '-C'\n"));
+			usage(B_FALSE);
+		}
+		if (projid != ZFS_INVALID_PROJID) {
+			(void) fprintf(stderr,
+			    gettext("'-p' is useless together with '-C'\n"));
+			usage(B_FALSE);
+		}
+		break;
+	case ZFS_PROJECT_SET:
+		if (!newline) {
+			(void) fprintf(stderr,
+			    gettext("'-0' is useless together with '-s'\n"));
+			usage(B_FALSE);
+		}
+		if (dironly) {
+			(void) fprintf(stderr,
+			    gettext("'-d' is useless together with '-s'\n"));
+			usage(B_FALSE);
+		}
+		if (keep) {
+			(void) fprintf(stderr,
+			    gettext("'-k' is useless together with '-s'\n"));
+			usage(B_FALSE);
+		}
+		break;
+	default:
+		if (projid == ZFS_INVALID_PROJID) {
```

Comment:
If we consider 0 as valid Project ID as ext4 etc, please use another bool value to indicate whether projectID is set.

Reply:
"0" is a valid project ID for ZFS. Here, the "ZFS_INVALID_PROJID" is defined as "-1". Means you can run "zfs project -p 0 file".